exonum/src/proto/push_proto_files.sh

```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env bash
+
+# This script copies protobuf specification files from exonum crates and pushes them as a separate new branch to the
+# repo that could be used by various Exonum clients. The branch can then be used as a source for a pull request.
+#
+# This script is intended to be used by Exonum developers in order to update Exonum clients with stable changes in the
+# proto files. No need to execute this script for any intermediate changes.
+
+set -eu -o pipefail
+
+# prints a section header
+function header() {
+    local title=$1
+    local rest="========================================================================"
+    echo
+    echo "===[ ${title} ]${rest:${#title}}"
+    echo
+}
+
+# TODO: Substitute with a real repo
+PROTO_REPO_URI="https://github.com/skletsun/test_repo.git"
+PROTO_REPO_TMP_DIR="/tmp/_proto_repo_tmp"
+
+BRANCH_RPEFIX="update-proto"
+
+PROTO_ROOT_DIR=$(pwd)
+MAIN_PROTO_FILES_DIR=${PROTO_ROOT_DIR}/schema/exonum
+COMPONENTS_DIR=$(pwd)/../../../components
+
+# Clean temporary dir from the previous iteration if any
+rm -fR ${PROTO_REPO_TMP_DIR}
+
+header "CLONING REPO"
+
+# Checkout repo
+git clone ${PROTO_REPO_URI} ${PROTO_REPO_TMP_DIR}
+
+header "COPYING PROTO FILES"
+
+# Copy main files
+cp -v ${MAIN_PROTO_FILES_DIR}/blockchain.proto ${PROTO_REPO_TMP_DIR}
+cp -v ${MAIN_PROTO_FILES_DIR}/consensus.proto ${PROTO_REPO_TMP_DIR}
+cp -v ${MAIN_PROTO_FILES_DIR}/runtime.proto ${PROTO_REPO_TMP_DIR}
+# BitVec
+cp -v ${COMPONENTS_DIR}/proto/src/proto/common.proto ${PROTO_REPO_TMP_DIR}
+# Crypto stuff
+cp -v ${COMPONENTS_DIR}/crypto/src/proto/schema/types.proto ${PROTO_REPO_TMP_DIR}
+
+header "ADDING PROTO FILES TO THE REPO"
+
+cd ${PROTO_REPO_TMP_DIR}
+
+# Create a new branch for changes and push
+BRANCH_NAME=${BRANCH_RPEFIX}-$(date "+%Y.%m.%d-%H%M%S")
```

Comment:
Do we ever need minutes and seconds here? I'd rather expect to see a version number (supplied by the user of the script)

Reply:
Actually, the initial idea was to use that resulting number as something unique (akin to version) in order to be able to quickly create a branch without syncing to version numbers (and branch then can safely be deleted after merging the PR). What do we really need the exact version for? It even may remain unchanged between the revisions of exonum.

Reply:
I think that you can automatically add `option java_package` during this script execution.

Reply:
> you can automatically add option java_package during this script 

I don't think it is sensible to introduce this complexity in the shell script. As these messages are considered to be public, they must be used *as is* by all the other runtimes to ensure compatibility (Java, Python, C#) — and acquire all the options needed. If they don't, some will eventually end up with incompatible classes for `exonum.Hash` in different packages.

Reply:
Options do not affect the serialization algorithm.

Reply:
I think that each runtime should fix flaws in protobuf design by itself to keep proto files clear.

Reply:
> Options do not affect the serialization algorithm.

On the wire != source compatibility.

> to keep proto files clear.

To the point of being useless?

Reply:
Okay, to make it clear, let's take a look at the extreme:
If some client X of the Exonum (let's make him a cash-bringing client to discard an option to ignore) will have an esoteric backend implemented in his personal langauge Y which requires a custom Y-specific metadata in the proto file, would we add this metadata (important *only* for that client) to the core repository (meant for the wide audience)?

Reply:
>To the point of being useless?
I don't think that sarcasm can be helpful in the discussion.

Reply:
Let's consider your esoteric language when it becomes realistic.

The protobuf design requires the codegeneration options to be specified in the file. Whether @alekseysidorov likes it or not, that's the way it is designed — to have a single source of the data description and compatible generated code. If the core decides to use protobuf, then it has to follow its design principles and guidelines to make it usable. To (a) ensure on-the wire *and* source compatibility; (b) have easy and reliable configuration in the downstream projects, the source files must have options to enable codegeneration in the [target languages](https://github.com/protocolbuffers/protobuf/blob/5b232b8ecbce13286be09e703997e887ae0d464d/src/google/protobuf/timestamp.proto#L35-L41). Currently the only other target language is Java — hence the Java options.